rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,774 @@
+            *n ^= b;
+        }
+
+        nonce
+    }
+
+    /// See RFC 9180 §5.2 "Encryption and Decryption"[^0].
+    ///
+    /// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2
+    fn increment_seq_num(&mut self) -> Result<(), aws_lc_rs::error::Unspecified> {
+        // def Context<ROLE>.IncrementSeq():
+        //   if self.seq >= (1 << (8*Nn)) - 1:
+        //     raise MessageLimitReachedError
+        //   self.seq += 1
+
+        // Determine the maximum sequence number using the AEAD nonce's length in bits.
+        // Do this as an u128 to prevent overflowing.
+        let max_seq_num = (1u128 << (self.aead.nonce_len() * 8)) - 1;
+
+        // Promote the u32 sequence number to an u128 and compare against the maximum allowed
+        // sequence number.
+        if u128::from(self.seq_num) >= max_seq_num {
+            return Err(aws_lc_rs::error::Unspecified);
+        }
+
+        self.seq_num += 1;
+        Ok(())
+    }
+}
+
+impl NonceSequence for &mut KeySchedule {
+    fn advance(&mut self) -> Result<AwsNonce, aws_lc_rs::error::Unspecified> {
+        let nonce = self.compute_nonce();
+        self.increment_seq_num()?;
+        AwsNonce::try_assume_unique_for_key(&nonce)
+    }
+}
+
+/// See RFC 9180 §4 "Cryptographic Dependencies"[^0].
+///
+/// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-4
+fn labeled_extract(
+    hkdf: &'static dyn Hkdf,
+    suite_id: &[u8],
+    salt: Option<&[u8]>,
+    label: &[u8],
+    ikm: &[u8],
+) -> Box<dyn HkdfExpander> {
+    // def LabeledExtract(salt, label, ikm):
+    //   labeled_ikm = concat("HPKE-v1", suite_id, label, ikm)
+    //   return Extract(salt, labeled_ikm)
+
+    let labeled_ikm = [&b"HPKE-v1"[..], suite_id, label, ikm].concat();
+    hkdf.extract_from_secret(salt, &labeled_ikm)
+}
+
+/// See RFC 9180 §4 "Cryptographic Dependencies"[^0].
+///
+/// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-4
+fn labeled_expand(
+    suite_id: &[u8],
+    expander: Box<dyn HkdfExpander>,
+    label: &[u8],
+    kem_context: &[u8],
+    len: u16,
+) -> Result<Vec<u8>, OutputLengthError> {
+    // def LabeledExpand(prk, label, info, L):
+    //   labeled_info = concat(I2OSP(L, 2), "HPKE-v1", suite_id,
+    //                         label, info)
+    //   return Expand(prk, labeled_info, L)
+
+    let output_len = u16::to_be_bytes(len);
+    let info = &[&output_len[..], b"HPKE-v1", suite_id, label, kem_context];
+
+    // TODO(XXX): consider reworking to use `crypto::tls13::expand()` with a compile size constant
+    //   output to make this infallible.
+    let mut expanded = vec![0; len as usize];
+    expander.expand_slice(info, &mut expanded)?;
+
+    Ok(expanded)
+}
+
+/// A newtype wrapper for a DH KEM shared secret.
+struct KemSharedSecret(Vec<u8>);
+
+fn expand_err(_: OutputLengthError) -> Error {
+    Error::General("HKDF expansion failed".into())
+}
+
+fn unspecified_err(_e: aws_lc_rs::error::Unspecified) -> Error {
```

Comment:
I think this could be a `feature`-guarded `From` impl for `Error`? Same for the other one below? That should work with `?` such that you don't need the `map_err()`s all over the place.

Reply:
Good idea.